pyogrio/geopandas.py

```diff
@@ -627,8 +729,33 @@ def write_dataframe(
             df = pd.DataFrame(df, copy=False)
             df[geometry_column] = geometry
 
+        # Arrow doesn't support datetime columns with mixed timezones, and GDAL only
+        # supports timezone offsets. Hence, to avoid data loss, convert columns that can
+        # contain datetime values with different offsets to strings.
+        # Also pass a list of these columns on so GDAL so it can still treat them as
+        # datetime columns when writing the dataset.
+        datetime_cols = []
+        for name, dtype in df.dtypes.items():
+            if dtype == "object":
+                # An object column with datetimes can contain multiple offsets.
+                if pd.api.types.infer_dtype(df[name]) == "datetime":
+                    df[name] = df[name].astype("string")
+                    datetime_cols.append(name)
+
+            elif isinstance(dtype, pd.DatetimeTZDtype) and str(dtype.tz) != "UTC":
+                # A pd.datetime64 column with a timezone different than UTC can contain
+                # data with different offsets because of summer/winter time.
+                df[name] = df[name].astype("string")
+                datetime_cols.append(name)
+
         table = pa.Table.from_pandas(df, preserve_index=False)
 
+        # Add metadata to datetime columns so GDAL knows they are datetimes.
+        for datetime_col in datetime_cols:
+            table = _add_column_metadata(
+                table, column_metadata={datetime_col: {"GDAL:OGR:type": "DateTime"}}
+            )
```

Comment:
```suggestion
        table = _add_column_metadata(
            table, column_metadata={col: {"GDAL:OGR:type": "DateTime"} for col in datetime_cols}
        )
```

?

The helper function seems to support multiple keys at once